edge/pkg/devicetwin/dtmanager/dmiserver.go

```diff
@@ -0,0 +1,361 @@
+	}
+	_, err = dc.RemoveDevice(ctx, cdr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func GrpcUpdateDevice(device *v1alpha2.Device, model *v1alpha2.DeviceModel) error {
+	protocol, err := dtcommon.GetProtocolNameOfDevice(device)
+	if err != nil {
+		return err
+	}
+	sockPath := GetSockPath(protocol)
+	if sockPath == "" {
+		return fmt.Errorf("cannot get sockPath of protocol %s", protocol)
+	}
+
+	dc, ctx, conn, cancelFunc, err := dmiclient.GenerateDMIClient(sockPath)
+	if err != nil {
+		return err
+	}
+
+	defer conn.Close()
+	defer cancelFunc()
+
+	cdr, err := dmiclient.UpdateDeviceRequest(device, model)
+	if err != nil {
+		return fmt.Errorf("fail to generate UpdateDeviceRequest for device %s with err: %v", device.Name, err)
+	}
+	_, err = dc.UpdateDevice(ctx, cdr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func GetSockPath(protocol string) string {
+	for _, mapper := range MapperList {
+		if mapper.GetProtocol() == protocol {
+			return string(mapper.Address)
+		}
+	}
+	return ""
+}
+
+func initDeviceModelInfoFromDB() {
+	DeviceModelList = make(map[string]*v1alpha2.DeviceModel)
+	metas, err := dao.QueryMeta("type", ResourceTypeDeviceModel)
+	if err != nil {
+		klog.Errorf("fail to init device model info from db with err: %v", err)
+		return
+	}
+
+	for _, meta := range *metas {
+		deviceModel := &v1alpha2.DeviceModel{}
+		if err := json.Unmarshal([]byte(meta), &deviceModel); err != nil {
+			klog.Errorf("fail to unmarshal device model info from db with err: %v", err)
+			return
+		}
+		klog.Infof("deviceModel: %+v", deviceModel)
+		DeviceModelList[deviceModel.Name] = deviceModel
+	}
+	klog.Infoln("success to init device model info from db")
+}
+
+func initPendingDeviceList() {
+	PendingDeviceList = make(map[string]*v1alpha2.Device)
+}
+
+func initDeviceInfoFromDB() {
+	DeviceList = make(map[string]*v1alpha2.Device)
+	metas, err := dao.QueryMeta("type", ResourceTypeDevice)
+	if err != nil {
+		klog.Errorf("fail to init device info from db with err: %v", err)
+		return
+	}
+
+	for _, meta := range *metas {
+		device := &v1alpha2.Device{}
+		if err := json.Unmarshal([]byte(meta), &device); err != nil {
+			klog.Errorf("fail to unmarshal device info from db with err: %v", err)
+			return
+		}
+		DeviceList[device.Name] = device
+	}
+	klog.Infoln("success to init device info from db")
+}
+
+func initDeviceMapperInfoFromDB() {
+	MapperList = make(map[string]*pb.MapperInfo)
+	metas, err := dao.QueryMeta("type", ResourceTypeDeviceMapper)
+	if err != nil {
+		klog.Errorf("fail to init device mapper info from db with err: %v", err)
+		return
+	}
+
+	for _, meta := range *metas {
+		deviceMapper := &pb.MapperInfo{}
+		if err := json.Unmarshal([]byte(meta), &deviceMapper); err != nil {
+			klog.Errorf("fail to unmarshal device mapper info from db with err: %v", err)
+			return
+		}
+		MapperList[deviceMapper.Name] = deviceMapper
+	}
+	klog.Infoln("success to init device mapper info from db")
+}
+
+func dealPendingDeviceList() {
```

Comment:
Do we need to poll periodically instead of doing it all the time?

Reply:
i have split the create device function to two funcs. thus the pending device list is no longer needed. i have removed it.